tests/ui/collapsible_span_lint_calls.rs

```diff
@@ -0,0 +1,93 @@
+// run-rustfix
+#![deny(clippy::internal)]
+#![feature(rustc_private)]
+
+extern crate rustc_ast;
+extern crate rustc_errors;
+extern crate rustc_lint;
+extern crate rustc_session;
+extern crate rustc_span;
+
+use rustc_ast::ast::Expr;
+use rustc_errors::{Applicability, DiagnosticBuilder};
+use rustc_lint::{EarlyContext, EarlyLintPass, Lint, LintContext};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::source_map::Span;
+
+#[allow(unused_variables)]
+pub fn span_lint_and_then<'a, T: LintContext, F>(cx: &'a T, lint: &'static Lint, sp: Span, msg: &str, f: F)
+where
+    F: for<'b> FnOnce(&mut DiagnosticBuilder<'b>),
+{
+}
+
+#[allow(unused_variables)]
+fn span_lint_and_help<'a, T: LintContext>(cx: &'a T, lint: &'static Lint, span: Span, msg: &str, help: &str) {}
+
+#[allow(unused_variables)]
+fn span_lint_and_note<'a, T: LintContext>(
+    cx: &'a T,
+    lint: &'static Lint,
+    span: Span,
+    msg: &str,
+    note_span: Span,
+    note: &str,
+) {
+}
+
+#[allow(unused_variables)]
+fn span_lint_and_sugg<'a, T: LintContext>(
+    cx: &'a T,
+    lint: &'static Lint,
+    sp: Span,
+    msg: &str,
+    help: &str,
+    sugg: String,
+    applicability: Applicability,
+) {
+}
+
+declare_tool_lint! {
+    pub clippy::TEST_LINT,
+    Warn,
+    "",
+    report_in_external_macro: true
+}
+
+declare_lint_pass!(Pass => [TEST_LINT]);
+
+impl EarlyLintPass for Pass {
+    fn check_expr(&mut self, cx: &EarlyContext, expr: &Expr) {
+        let lint_msg = "lint message";
+        let help_msg = "help message";
+        let note_msg = "note message";
+        let sugg = "new_call()";
+        let predicate = true;
+
+        span_lint_and_then(cx, TEST_LINT, expr.span, lint_msg, |db| {
+            db.span_suggestion(expr.span, help_msg, sugg.to_string(), Applicability::MachineApplicable);
+        });
+        span_lint_and_then(cx, TEST_LINT, expr.span, lint_msg, |db| {
+            db.span_help(expr.span, help_msg);
```

Comment:
This should not get linted, looking at how the wrapper functions are implemented. But since it makes sense to lint this, I would suggest to adapt the wrapper functions, instead of the lint.

Currently the wrapper functions `span_lint_and_note` and `span_lint_and_help` are inconsitent, since `note` takes an additional span, while `help` doesn't. I'd suggest to make following two changes:

1. Turn the `note_span` argument in `span_lint_and_note` to an `Option<Span>` to make it more idomatic
2. Add a `help_span: Option<Span>` to `span_lint_and_help` and adapt the implementation to the one of `span_lint_and_note`

Reply:
@flip1995 Do you mean change the `help_span` argument in the ways that point 1 describe when you are saying 'adapt the implementation to the one of span_lint_and_note'?

Reply:
Yes, this will kind of conflict with my explanation above, for what to use wrapper functions, since every call to the wrapper will have the same `None` argument. But this is more clean, since help messages can have different spans from the error message and we should respect that in the future, like we do with `span_lint_and_note`.

After this PR, _someone_ can look over every `span_lint_and_help` call and determine, if providing another span would be helpful.

Reply:
@xiongmao86 Sorry if this is more work, than you might have expected. But implementing internal lints usually uncovers flaws in the code base, which should be fixed directly. Thanks for putting all this work in!

Reply:
Just wondering there is opportunity to reduce work. If there is tricks to do, it's good. If there is no tricks, I am still glad that I can help.

Reply:
There's not really a trick on how to fix the fallout from this change. In vim, I would record a macro, that adds the argument and then just `rg` for `span_help_and_lint` and jump between all occurrences.

Reply:
I have finish span_lint_and_help last night, I have to say it not so much work than I feel at the beginning. I'll add this tip to my note, thanks for the tip.

Reply:
Is `rg` a plug-in? `vim-rg`?

Reply:
Macro is not very fit solution, does rust plan to have refactoring tool?

Reply:
> does rust plan to have refactoring tool?

Rust has rustfix. But this doesn't really work for missing arguments, since automatically add a random argument will almost always be the wrong suggestion. I don't know how well rustfix is integrated with vscode yet.

Reply:
Thanks for the tips. I have adapted the `span_lint_and_note` invocation. Waiting for toolchain updating.